### PR TITLE
go/libraries/events: Restructure events collection so that we can send events while the dolt process is running.

### PR DESCRIPTION
**Pull request by reltuk**
We make the GlobalCollector mutable.

We give the Collector a sendingThread which can have an Emitter. When the Collector collects enough events, it forwards a batch of them to the sendingThread. If the sendingThread has an Emitter, it attempts to send its current batch on that Emitter.

On shutdown, contexts get canceled, background threads get waited on, and then an accumulated un-emitted events are returned from the Collector, same as previously.

### CODE REVIEW

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999562 to 99.999562</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>9a3ceef</TD>
<TD>not ok</TD>
<TD>26</TD>
</TR>
<TR><TD>9a3ceef</TD>
<TD>ok</TD>
<TD>5937431</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>9a3ceef</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999562</TD>
</TR>
</table>

**Comment by coffeegoddd** [comment]
@reltuk __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999562 to 99.999562</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>3902282</TD>
<TD>not ok</TD>
<TD>26</TD>
</TR>
<TR><TD>3902282</TD>
<TD>ok</TD>
<TD>5937431</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>3902282</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999562</TD>
</TR>
</table>

File: go/libraries/events/collector.go [hunk 1]
```diff
@@ -40,34 +42,59 @@ func getMachineID() string {
...
+	cur := globalCollector
+	globalCollector = c
+	toTransfer := cur.Close()
+	for _, e := range toTransfer {
+		globalCollector.evtCh <- e
+	}
+}
 
 const collChanBufferSize = 32
+const maxBatchedEvents = 1024
```
Comment (zachmu) [inline go/libraries/events/collector.go:66]: Any reason to set this so high? Seems like less bursty would be better

File: go/libraries/events/collector.go [hunk 2]
```diff
@@ -85,7 +112,89 @@ func (c *Collector) Close() []*eventsapi.ClientEvent {
...
-	interf := c.val.Load()
+	return c.events
+}
+
+type sendingThread struct {
+	logCtx  context.Context
+	cancelF func()
+
+	batchCh chan []*eventsapi.ClientEvent
+	unsent  []*eventsapi.ClientEvent
```
Comment (zachmu) [inline go/libraries/events/collector.go:123]: Seems like this is actually a local var?

**Review by zachmu — APPROVED** [review]
LGTM!

File: go/libraries/events/collector.go (hunk 2 above)
Comment (reltuk) [reply go/libraries/events/collector.go:123]: Almost. Returned from `stop()` for any un-emitted things.

**Comment by coffeegoddd** [comment]
@coffeegoddd __DOLT__
 <table>
<TR><TH>comparing_percentages</TH>
</TR>
<TR><TD>99.999562 to 99.999562</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>result</TH>
<TH>total</TH>
</TR>
<TR><TD>82a20f2</TD>
<TD>not ok</TD>
<TD>26</TD>
</TR>
<TR><TD>82a20f2</TD>
<TD>ok</TD>
<TD>5937431</TD>
</TR>
</table>
<table>
<TR><TH>version</TH>
<TH>total_tests</TH>
</TR>
<TR><TD>82a20f2</TD>
<TD>5937457</TD>
</TR>
</table>
<table>
<TR><TH>correctness_percentage</TH>
</TR>
<TR><TD>99.999562</TD>
</TR>
</table>

**Pull request merged**